packages/x-data-grid/src/hooks/features/rowSelection/utils.ts

```diff
@@ -13,3 +17,106 @@ export function isMultipleRowSelectionEnabled(
   }
   return !props.disableMultipleRowSelection;
 }
+
+const getRowNodeParents = (tree: GridRowTreeConfig, id: GridRowId) => {
+  const parents: GridRowId[] = [];
+
+  let parent: GridRowId | null = id;
+
+  while (parent && parent !== GRID_ROOT_GROUP_ID) {
+    const node = tree[parent] as GridGroupNode;
+    if (!node) {
+      return parents;
+    }
+    parents.push(parent);
+    parent = node.parent;
+  }
+  return parents;
+};
+
+const getFilteredRowNodeSiblings = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  tree: GridRowTreeConfig,
+  filteredRows: Record<GridRowId, boolean>,
+  id: GridRowId,
+) => {
+  const node = apiRef.current.getRowNode(id);
+  if (!node) {
+    return [];
+  }
+
+  const parent = node.parent;
+  if (!parent) {
+    return [];
+  }
+
+  const parentNode = tree[parent] as GridGroupNode;
+
+  return [...parentNode.children].filter((childId) => childId !== id && filteredRows[childId]);
+};
+
+export const findRowsToSelect = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  tree: GridRowTreeConfig,
+  selectedRow: GridRowId,
+) => {
+  const filteredRows = gridFilteredRowsLookupSelector(apiRef);
+  const rowsToSelect: GridRowId[] = [];
+
+  const traverseParents = (rowId: GridRowId) => {
+    const siblings: GridRowId[] = getFilteredRowNodeSiblings(apiRef, tree, filteredRows, rowId);
+    if (
+      siblings.length === 0 ||
+      siblings.every((sibling) => apiRef.current.isRowSelected(sibling))
+    ) {
+      const rowNode = apiRef.current.getRowNode(rowId) as GridGroupNode;
+      const parent = rowNode.parent;
+      if (parent && parent !== GRID_ROOT_GROUP_ID) {
+        rowsToSelect.push(parent);
+        traverseParents(parent);
+      }
+    }
+  };
+  traverseParents(selectedRow);
+
+  const rowNode = apiRef.current.getRowNode(selectedRow);
+
+  if (rowNode?.type === 'group') {
+    const children = apiRef.current.getRowGroupChildren({
+      groupId: selectedRow,
+      skipAutoGeneratedRows: false,
+      applyFiltering: true,
+    });
+    children.forEach((child) => rowsToSelect.push(child));
+  }
+  return rowsToSelect;
+};
+
+export const findRowsToDeselect = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  tree: GridRowTreeConfig,
+  deselectedRow: GridRowId,
+) => {
+  let rowsToDeselect: GridRowId[] = [];
+
+  const allParents = getRowNodeParents(tree, deselectedRow);
+  allParents.forEach((parent) => {
+    const isSelected = apiRef.current.isRowSelected(parent);
+    if (isSelected) {
+      rowsToDeselect.push(parent);
+    }
+  });
+
+  const rowNode = apiRef.current.getRowNode(deselectedRow);
+  if (rowNode?.type === 'group') {
+    rowsToDeselect = [
+      ...rowsToDeselect,
+      ...apiRef.current.getRowGroupChildren({
```

Comment:
would this be fully recursive?

Reply:
It uses `.children` property of each group node to recursively find all the children.